### ISSUE TITLE
Textures not rendered properly when using multi-page texture atlas

### ISSUE THREAD

**Issue by dariuspranskus** [CypherCove/FlexBatch#1]
Hello

I am using CompliantBatch to render all my game actors and everything is fine when I map textureRegions from a single page atlas to my textures to be sent to my shader. But if I need to pack more textures and my atlas becomes multi-page, textures from the second page are not rendered properly as if the textcoords were not correct. 

So basically what I do is:

- load texture atlas using AssetManager
- set number of textures to 5 in my extended Quad2D class
- set textures using textureRegion(atlas.findRegion("...")) function calls. For some objects one texture region might come from one atlas page and another from different page.
- use u_texture* and a_texCoords* (passed from vertex shader to fragment shader as variant) to access sampler2D and UVs of the textures.

What could be a problem? Am I missing something?

Thanks

**Comment by cypherdare** [comment]
Now that you're describing it, I don't think I ever tested this condition--combining multi-texturing and mutlitple TextureRegions from different Texture sources, and then mixing and matching them. If it were working properly, I think it should detect when one batchable's five textures don't match the five textures from the previous one, so it should flush them before queuing the next. 

I'll try reproduce this in the next day or two and see what the problem is.

That said, even with this working, it sounds like you will have a lot of texture swapping and draw calls going on. I would recommend trying to organize the atlas pages.

**Comment by dariuspranskus** [comment]
Thanks. Actually most of the space in my textures is the whitespace, which I would be happy to trim, but I am laying multiple textures in my shader. For example I may render the object which has a body and an icon, and the icon is much smaller than the body. I haven't figured it out how to do this with the textures of different size, currently if I remove whitespace, it enlarges that texture when rendering. SO I went with the approach to have all my textures the same size with the transparent pixels for the whitespace, but this takes a lot of space in my atlas.

**Comment by cypherdare** [comment]
OK, I've reproduced the issue and I see two contributing factors. I'll try to put together a fix for you to try out in the next couple of days.

**Comment by cypherdare** [comment]
I think this hack should work for the time being. Subclass CompliantBatch, and add this method:
```
@Override
protected void applyMatrices () {
    super.applyMatrices();
    final boolean wasPendantic = ShaderProgram.pedantic;
    ShaderProgram.pedantic = false;
    applyTextureUniforms();
    ShaderProgram.pedantic = wasPendantic;
}
```

**Comment by dariuspranskus** [comment]
Thank you @cypherdare. I also tried to have one bigger atlas page, 4096 x 2048 in size, not sure if it is a good idea and how it will be supported by some older devices. Probably the good idea would be use that single large texture, if supported by hardware and switch to  workaround above, if not.

**Comment by cypherdare** [comment]
Resolved with commit 08c671612690cd04c072b735cdfb956c491c1a41.

I released a new version 1.2.2 with the fix, so the hack is no longer necessary if you update the library.

If you're targeting Android, some (many?) devices only support texture sizes up to 2048x2048. IMO, managing large and small versions of the atlas is not a great solution. The devices that cannot support the larger size are also the ones more likely to suffer from frequent texture swapping anyway.

It might be possible to create a Batchable that takes uses AtlasRegions instead of TextureRegions and reverses the whitespace stripping in the fragment shader. However, this would require 8 more bytes per texture per vertex in the vertex data, to carry two more floats that determine how much area to make invisible. And it would be kind of complicated to write.

Another option might be to write a version of TextureAtlas that loads the pages into a single TextureArray instead of multiple Textures. Then there would be no texture swapping at all. But this would also be complicated to write. I have a pull request for LibGDX that would create this type of TextureAtlas, but it's been sitting in limbo for a couple years. Maybe I could adapt part of that into the FlexBatch library. I might first try breaking it up into smaller incremental PRs to see how much groundwork can be put in LibGDX first.

So maybe the easiest solution right now would be to simple draw your five separate TextureRegions on top of each other as separate items. If you use AtlasRegion, it has data for the offset due to whitespace stripping, so you could use stripped whitespace and keep your atlas smaller.

**Comment by dariuspranskus** [comment]
Thank you @cypherdare.
